### PR TITLE
Add Prometheus Monitoring Service

### PR DESCRIPTION
**Pull request by YanjieGao**
(1) Prometheus Monitoring service module
(2) This part has been unit test at east us cluster
(3) README introduction:
# Prometheus Monitoring Service
The monitor service is the entrance for cluster monitoring.
User can monitor cluster and node through the web UI.

## Installation
The [readme](../service-deployment/README.md) in service deployment introduces the overall installation process, including that of the web portal. 
The following parameters in the [clusterconfig.yaml](../service-deployment/clusterconfig-example.yaml) are of interest to web portal.
* prometheus_addr: String, the address of the prometheus dashboard, for example, http://10.0.3.9:9090
* prometheus_port: Int, the port to use when launching prometheus, for example, 9090
* grafana_port: Int, the port to use when launching node exporter, for example, 9100

### View cluster status 

Click the link (For example: http://10.0.3.9:9090) to go to the dashboard of prometheus, where it shows all nodes metrics running on the cluster.

### CODE REVIEW

File: service-deployment/bootstrap/prometheus/start.sh [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+# (1) create configmap
+eval "${KUBECTL} create -f prometheus-configmap.yaml"
+eval "${KUBECTL} create configmap grafana-import-dashboards --from-file=dashboards -o json --dry-run" | eval "${KUBECTL} apply -f -"
+
+# (2) create the pods
+eval "${KUBECTL} create -f gpu-exporter-ds.yaml"
+eval "${KUBECTL} create -f node-exporter-ds.yaml"
+eval "${KUBECTL} create -f prometheus-deployment.yaml"
+eval "${KUBECTL} create -f grafana.yaml"
+eval "${KUBECTL} create -f grafana-cmd.yaml"
```
[the remark this answers is not in the document]
Comment (YanjieGao) [reply service-deployment/bootstrap/prometheus/start.sh]: solved, remove to next PR

File: service-deployment/bootstrap/prometheus/grafana.yaml.template [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+          value: grafana-simple-json-datasource
+        volumeMounts:
+          - mountPath: /usr/share/grafana/public/dashboards
+            name: config-volume
+      volumes:
+         - name: config-volume
+           configMap:
+             name: grafana-import-dashboards
+         - name: grafana-persistent-storage
+           emptyDir: {}
```
[the remark this answers is not in the document]
Comment (YanjieGao) [reply service-deployment/bootstrap/prometheus/grafana.yaml.template]: solved removed

File: service-deployment/bootstrap/prometheus/grafana-cmd.yaml.template [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            done
+        volumeMounts:
+        - name: config-volume
+          mountPath: /opt/grafana-import-dashboards
+      volumes:
+         - name: config-volume
+           configMap:
+             name: grafana-import-dashboards
+         - name: grafana-persistent-storage
+           emptyDir: {}
```
[the remark this answers is not in the document]
Comment (YanjieGao) [reply service-deployment/bootstrap/prometheus/grafana-cmd.yaml.template]: solved removed

File: service-deployment/bootstrap/prometheus/grafana-cmd.yaml.template [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      hostNetwork: true
+      hostPID: true
+      nodeSelector:
+        prom: prom
+      containers:
+      - name: grafana-import-dashboards
+        image: docker
+        command: ["/bin/sh", "-c"]
+        workingDir: /opt/grafana-import-dashboards
+        args:
```
[the remark this answers is not in the document]
Comment (YanjieGao) [reply service-deployment/bootstrap/prometheus/grafana-cmd.yaml.template]: removed to next PR

File: service-deployment/bootstrap/prometheus/grafana-cmd.yaml.template [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        task: monitoring
+        k8s-app: grafana
+    spec:
+      hostNetwork: true
+      hostPID: true
+      nodeSelector:
+        prom: prom
+      containers:
+      - name: grafana-import-dashboards
+        image: docker
```
[the remark this answers is not in the document]
Comment (YanjieGao) [reply service-deployment/bootstrap/prometheus/grafana-cmd.yaml.template]: solved

File: service-deployment/bootstrap/prometheus/gpu-exporter-ds.yaml.template [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  name: gpu-exporter
+spec:
+  template:
+    metadata:
+      labels:
+        app: gpu-exporter
+      name: gpu-exporter
+    spec:
+      containers:
+      - image: prom11/gpu_exporter:v14
```
[the remark this answers is not in the document]
Comment (YanjieGao) [reply service-deployment/bootstrap/prometheus/gpu-exporter-ds.yaml.template]: solved removed

File: service-deployment/src/gpu-exporter/dockerfile.template [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+ENV LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$NV_DRIVER/lib:$NV_DRIVER/lib64
+ENV PATH=$PATH:$NV_DRIVER/bin
+
+RUN apt-get update && \
+    apt-get -y install golang --no-install-recommends && \
+    rm -r /var/lib/apt/lists/*
+
+WORKDIR /go
+
+COPY . .
```
[the remark this answers is not in the document]
Comment (YanjieGao) [reply service-deployment/src/gpu-exporter/dockerfile.template]: solved removed

File: service-deployment/src/gpu-exporter/dockerfile.template [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+# BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+FROM nvidia/cuda
+
+ENV NVIDIA_VERSION=375.26
```
[the remark this answers is not in the document]
Comment (YanjieGao) [reply service-deployment/src/gpu-exporter/dockerfile.template]: ask yundong, change to current

[the remark this answers is not in the document]
Comment (YanjieGao) [reply service-deployment/src/gpu-exporter/dockerfile.template]: solved

File: service-deployment/src/gpu-exporter/dockerfile [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+# to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+# BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+FROM nvidia/cuda
```
Comment (ydye) [inline service-deployment/src/gpu-exporter/dockerfile]: here, maybe fix the version tag will be better?

Comment (YanjieGao) [reply service-deployment/src/gpu-exporter/dockerfile]: solved

File: service-deployment/service.yaml [hunk 10]
```diff
@@ -72,6 +72,11 @@ imagelist:
       - dockerfile
     prerequisite: None
 
+  gpu-exporter:
+    templatelist:
+      - dockerfile
+    prerequisite: drivers
```
Comment (ydye) [inline service-deployment/service.yaml]: Why gpu-exporter depends on drivers image? Here is the image list, not service list. If your image doesn't depend on drivers, you should set None here.

Comment (YanjieGao) [reply service-deployment/service.yaml]: Solved #Resolved

File: service-deployment/src/gpu-exporter/dockerfile (hunk 9 above)
Comment (fanyangCS) [reply service-deployment/src/gpu-exporter/dockerfile]: @YundongYe@outlook.com I remember you will always install the latest driver in driver docker. should we make the two consistent?

---
In reply to: [157393359](https://github.com/Microsoft/pai/pull/87#discussion_r157393359) [](ancestors = 157393359)

File: service-deployment/service.yaml [hunk 11]
```diff
@@ -72,6 +72,11 @@ imagelist:
       - dockerfile
     prerequisite: None
 
+  gpu-exporter:
```
Comment (fanyangCS) [inline service-deployment/service.yaml]: >exporter [](start = 6, length = 8)

gpu exporter always pre-require node exporter, right?

Comment (YanjieGao) [reply service-deployment/service.yaml]: No they have no dependency

**Pull request closed without merging**